### PR TITLE
Better performing string replacements

### PR DESCRIPTION
**Pull request by brejoc**
Previously we where iterating over the whole search table to do string replacements. This can and will in most of the cases be very wasteful.

In order to prevent this, a regex lookup is now performed that returns the keys that are used in the template and only for those keys we do a str.replace().

Here are some tests with 60 iterations of `cobbler profile create …` and `cobbler profile remove`.

```
##############################
#    Without regex lookup    #
##############################

localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m28.936s
user    0m47.902s
sys     0m3.037s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m16.252s
user    0m47.524s
sys     0m3.197s
localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m28.280s
user    0m47.958s
sys     0m2.999s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m16.329s
user    0m47.646s
sys     0m3.184s


##############################
#      With regex lookup     #
##############################

localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m20.998s
user    0m47.726s
sys     0m3.298s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m11.269s
user    0m47.901s
sys     0m3.043s
localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m21.509s
user    0m48.258s
sys     0m3.061s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m11.319s
user    0m47.692s
sys     0m3.206s
```

### CODE REVIEW

**Comment by brejoc** [comment]
I think there is still a problem with this. When there is a key occurring multiple times in the template, we'd do unnecessary replacements.

**Comment by brejoc** [comment]
Last commit addresses the unnecessary replacements and the performance profile stays pretty much the same. Switching this from draft to regular PR.

```
localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m21.624s
user    0m48.327s
sys     0m2.961s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m11.895s
user    0m47.948s
sys     0m3.110s
localhost:/home/brejoc # time ./fake-cobbler.sh create
real    1m20.993s
user    0m47.904s
sys     0m2.968s
localhost:/home/brejoc # time ./fake-cobbler.sh remove
real    1m12.033s
user    0m48.045s
sys     0m3.083s
```

File: cobbler/templar.py [hunk 1]
```diff
@@ -139,9 +140,12 @@ def render(self, data_input, search_table, out_path, subject=None, template_type
             repstr = server
         search_table["http_server"] = repstr
 
-        for x in list(search_table.keys()):
-            if type(x) == str:
-                data_out = data_out.replace("@@%s@@" % str(x), str(search_table[str(x)]))
+        # string replacements for @@xyz@@ in data_out with prior regex lookups of keys
+        regex = r"@@[a-zA-Z]*@@"
```
Comment (SchoolGuy) [inline cobbler/templar.py:144]: I think it is possible to include also stuff like numbers etc. in the variable names. If we don't want to support this I am fine with it but we would need to document this regression in the docs!

Comment (brejoc) [reply cobbler/templar.py:144]: That could very well be, but I'm not sure about it. I haven seen a key with a number yet.

**Comment by watologo1** [comment]
I wonder why or in which (special) case this is needed?

instead of defining @@special_key@@
template magic should be used to replace things?
I do not understand the comment above:
        # Now apply some magic post-filtering that is used by "cobbler import" and some other places. Forcing folks to                                                     
        # double escape things would be very unwelcome.
Ideally the whole self made @@xxx@@ replacing would be nice to get rid of?
If possibly and I have idea on a quick glance whether this is the case.

BTW: Are there efforts to get rid of Cheetah in favor of Jinja?
Afaik Jinja is more widespread, etc.?
Maybe we should at least switch to default Jinja2 in latest head if this is the case?

File: cobbler/templar.py (hunk 1 above)
Comment (brejoc) [reply cobbler/templar.py]: Regex is more greedy now. I've switched to `r"@@[\S]*@@"`.

**Comment by SchoolGuy** [comment]
@brejoc Could we do a force push with an empty or whitespace commit so we retrigger Travis. We recently go moved to .com from .org and I am not sure if this either breakes the tests or not.

**Comment by SchoolGuy** [comment]
If the tests are broken I know why and this is easily fixable. I will then post the patch required to fix them for you.

**Comment by brejoc** [comment]
@SchoolGuy I'm just running the test locally to be sure.

**Comment by SchoolGuy** [comment]
The patch for fixing the broken tests can be found at: https://github.com/cobbler/cobbler/commit/c292fc7d7e0e7d1a0f67d22541b496e631afa24e

**Comment by brejoc** [comment]
@SchoolGuy Thx! Mind if I just cherr-pick it into this PR? Seems to not be in any branch yet.

**Comment by SchoolGuy** [comment]
@brejoc Do so. Then I will remove it from my PR.

**Comment by brejoc** [comment]
@SchoolGuy Not sure if 0b70da0c32a79408e613f20943f293c453b8515c make sense to you. Just let me know!

**Review by SchoolGuy — APPROVED** [review]
This is a LGTM PR for me. Thanks for the performance issue improvements!

**Review by hbokh — APPROVED** [review]
LGTM

**Comment by SchoolGuy** [comment]
Travis passes. Two approvals. Let's get this into `master`!

**Pull request merged**